### PR TITLE
Use GH Action instead of `on` path filtering

### PR DESCRIPTION
**Pull request by kenodegard**
### Description

We've been struggling to get our tests.yml and tests-skip.yml workflows to work nicely together:

- rebases & merge updates sometimes resulted in the skip workflow being run even though there were code changes
- committing actual code changes sometimes would run the skip workflow only and other times (if the concurrency groups were different) both workflows would run

In either case, the branch protection required checks for merging would be satisfied before the tests were completed failing to prevent failing PRs from being merged.

This change merges the tests and skip workflows back together and uses a GitHub Action to detect whether changes have been made to code files. We then use `if` clauses to decide whether any jobs are being skipped.

Xref https://github.com/conda/conda/issues/12133 https://github.com/conda/conda-build/pull/4675

### Checklist - did you ...

- [ ] ~Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?~
     
- [ ] ~Add / update necessary tests?~
- [ ] ~Add / update outdated documentation?~

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -39,7 +25,36 @@ concurrency:
...
+    permissions:
+      pull-requests: read
+    outputs:
+      code: ${{ steps.filter.outputs.code }}
+    steps:
+      - uses: actions/checkout@v3
+        # dorny/paths-filter needs git clone for push events
+        # https://github.com/marketplace/actions/paths-changes-filter#supported-workflows
+        if: github.event_name == 'push'
+      - uses: dorny/paths-filter@v2
```
Comment (jezdez) [inline .github/workflows/tests.yml:40]: We should pin to a commit here or fork the action into conda/actions.
```suggestion
      - uses: dorny/paths-filter@4512585405083f25c027a35db413c2b3b9006d50
```

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -39,7 +25,36 @@ concurrency:
...
+        id: filter
+        with:
+          filters: |
+            code:
+              - 'conda/**'
+              - 'conda_env/**'
+              - 'tests/**'
+              - '*.py'
```
Comment (jezdez) [inline .github/workflows/tests.yml:48]: There is Python code in the `docs` folder FWIW. Probably not a huge deal, but just wanted to mention it.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -280,10 +310,12 @@ jobs:
...
 
-  analyze:
-    name: Analyze results
-    needs: [windows, linux, linux-qemu, macos]
-    if: always()
+  # aggregate and upload
+  aggregate:
+    # only aggregate test suite if there are code changes
+    needs: [changes, windows, linux, linux-qemu, macos]
+    if: needs.changes.outputs.code == 'true' && always()
```
Comment (jezdez) [inline .github/workflows/tests.yml:317]: That reads a little weird, do we need the `always()` here?

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -303,12 +335,21 @@ jobs:
...
+  # required check
+  analyze:
+    name: Analyze results
+    needs: [windows, linux, linux-qemu, macos, aggregate]
+    if: always()
+
+    runs-on: ubuntu-latest
+    steps:
       - name: Decide whether the needed jobs succeeded or failed
         uses: re-actors/alls-green@release/v1
```
Comment (jezdez) [inline .github/workflows/tests.yml:347]:
```suggestion
        uses: re-actors/alls-green@05ac9388f0aebcb5727afa17fcccfecd6f8ec5fe
```

**Review by jezdez — COMMENTED** [review]
This is smart, much easier to grok then the paths parameter.

**Review by jezdez — APPROVED** [review]

**Review by travishathaway — APPROVED** [review]

File: .github/workflows/tests.yml (hunk 3 above)
Comment (kenodegard) [reply .github/workflows/tests.yml:317]: Apparently, we do. The `always()` function does more than just evaluate to `true`. I had previously just used `if: needs.changes.outputs.code == 'true'` but that resulted in this job being skipped if any of the `needs` failed.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (kenodegard) [reply .github/workflows/tests.yml]: Currently, across both conda and conda-build we utilize a number of actions that we do not fork into conda/actions:

- actions-ecosystem/action-add-labels
- actions-ecosystem/action-remove-labels
- actions/add-to-project
- actions/cache
- actions/checkout
- actions/download-artifact
- actions/github-script
- actions/setup-python
- actions/stale
- actions/upload-artifact
- andstor/file-existence-action
- codecov/codecov-action
- conda-incubator/setup-miniconda
- conda/actions/canary-release
- conda/actions/check-cla
- conda/actions/read-yaml
- dessant/lock-threads
- docker/build-push-action
- docker/login-action
- docker/metadata-action
- docker/setup-buildx-action
- docker/setup-qemu-action
- dorny/paths-filter
- EndBug/label-sync
- re-actors/alls-green
- test-summary/action

Are we expected to fork all of these other actions as well? I would strongly resist any such effort.

That said, I can get on the commit hash bandwagon. Do we want to open a new issue to track updating our workflows to use commit hashes for their `uses`?

**Pull request merged**